der/src/document.rs

```diff
@@ -71,49 +106,99 @@ pub trait Document<'a>: AsRef<[u8]> + Sized + TryFrom<Vec<u8>, Error = Error> {
-            self.as_ref(),
+            self.as_der(),
         )?)
     }
 
     /// Read ASN.1 DER document from a file.
     #[cfg(feature = "std")]
     #[cfg_attr(docsrs, doc(cfg(feature = "std")))]
-    fn read_der_file(path: impl AsRef<Path>) -> Result<Self> {
+    pub fn read_der_file(path: impl AsRef<Path>) -> Result<Self>
+    where
+        T: for<'a> Decodable<'a> + Sized,
+    {
         fs::read(path)?.try_into()
     }
 
+    /// Write ASN.1 DER document to a file.
+    #[cfg(feature = "std")]
+    #[cfg_attr(docsrs, doc(cfg(feature = "std")))]
+    pub fn write_der_file(&self, path: impl AsRef<Path>) -> Result<()> {
+        write_file(path, self.as_der(), T::SENSITIVE)
+    }
+
     /// Read PEM-encoded ASN.1 DER document from a file.
     #[cfg(all(feature = "pem", feature = "std"))]
     #[cfg_attr(docsrs, doc(cfg(all(feature = "pem", feature = "std"))))]
-    fn read_pem_file(path: impl AsRef<Path>) -> Result<Self>
+    pub fn read_pem_file(path: impl AsRef<Path>) -> Result<Self>
     where
-        Self: pem::PemLabel,
+        T: for<'a> Decodable<'a> + PemLabel + Sized,
     {
         Self::from_pem(&fs::read_to_string(path)?)
     }
 
-    /// Write ASN.1 DER document to a file.
-    #[cfg(feature = "std")]
-    #[cfg_attr(docsrs, doc(cfg(feature = "std")))]
-    fn write_der_file(&self, path: impl AsRef<Path>) -> Result<()> {
-        write_file(path, self.as_ref(), Self::SENSITIVE)
-    }
-
     /// Write PEM-encoded ASN.1 DER document to a file.
     #[cfg(all(feature = "pem", feature = "std"))]
     #[cfg_attr(docsrs, doc(cfg(all(feature = "pem", feature = "std"))))]
-    fn write_pem_file(&self, path: impl AsRef<Path>, line_ending: pem::LineEnding) -> Result<()>
+    pub fn write_pem_file(&self, path: impl AsRef<Path>, line_ending: pem::LineEnding) -> Result<()>
     where
-        Self: pem::PemLabel,
+        T: PemLabel,
     {
-        write_file(path, self.to_pem(line_ending)?.as_bytes(), Self::SENSITIVE)
+        write_file(path, self.to_pem(line_ending)?.as_bytes(), T::SENSITIVE)
+    }
+
+    /// Attempt to decode `self.der_bytes` as `T`.
+    ///
+    /// This method doesn't uphold the invariant that `T` always decodes
+    /// successfully, but is needed to make the lifetimes for the constructor
+    /// work.
+    fn try_decode<'a>(&'a self) -> Result<T>
+    where
+        T: Decodable<'a> + Sized,
+    {
+        T::from_der(self.as_der())
+    }
+}
+
+impl<T> AsRef<[u8]> for Document<T>
+where
+    T: Sensitivity,
+{
+    fn as_ref(&self) -> &[u8] {
+        self.as_der()
+    }
+}
+
+// NOTE: `Drop` is defined unconditionally to ensure bounds do not change based
+// on selected cargo features, which would not be a purely additive change
+impl<T> Drop for Document<T>
+where
+    T: Sensitivity,
+{
+    fn drop(&mut self) {
+        #[cfg(feature = "zeroize")]
+        if T::SENSITIVE {
+            use zeroize::Zeroize;
+            self.der_bytes.zeroize();
+        }
+    }
+}
```

Comment:
It would probably be easier to just unconditionally `Zeroize<Vec<u8>>`. I presume you're worried about performance of zeroing large certificates. If you're willing to take the performance tradeoffs, zeroing even the non-sensitive bits has security benefits. For example, attackers often search for the public key to correlate the private key's memory position (or the position of some of its non-zeroed fragments). Zeroing both adds one additional hurdle.

Reply:
It's an option, but the sensitivity information is already there for other purposes, so it seems ok to me to make use of it for this application as well.